cpp/src/parquet/metadata_benchmark.cc

```diff
@@ -0,0 +1,152 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include <memory>
+#include <sstream>
+
+#include <benchmark/benchmark.h>
+
+#include "arrow/buffer.h"
+#include "arrow/io/memory.h"
+#include "arrow/util/logging.h"
+
+#include "parquet/column_writer.h"
+#include "parquet/file_reader.h"
+#include "parquet/file_writer.h"
+#include "parquet/metadata.h"
+#include "parquet/platform.h"
+#include "parquet/schema.h"
+
+namespace parquet {
+
+using ::arrow::Buffer;
+using ::arrow::io::BufferOutputStream;
+using ::arrow::io::BufferReader;
+using schema::GroupNode;
+using schema::NodePtr;
+using schema::NodeVector;
+
+class MetadataBenchmark {
+ public:
+  explicit MetadataBenchmark(benchmark::State* state)
+      : MetadataBenchmark(static_cast<int>(state->range(0)),
+                          static_cast<int>(state->range(1))) {}
+
+  MetadataBenchmark(int num_columns, int num_row_groups)
+      : num_columns_(num_columns), num_row_groups_(num_row_groups) {
+    NodeVector fields;
+    for (int i = 0; i < num_columns_; ++i) {
+      std::stringstream ss;
+      ss << "col" << i;
+      fields.push_back(parquet::schema::Int32(ss.str(), Repetition::REQUIRED));
+    }
+    schema_root_ = std::static_pointer_cast<GroupNode>(
+        GroupNode::Make("schema", Repetition::REQUIRED, fields));
+
+    WriterProperties::Builder prop_builder;
+    writer_properties_ =
+        prop_builder.version(ParquetVersion::PARQUET_2_6)->disable_dictionary()->build();
```

Comment:
Should we later add benchmark for drop statistics? ( this is useful in some case )

Reply:
(besides, should we later add benchmark for byte-array statistics decoding ?)

Reply:
What do you mean with "drop statistics"?

Reply:
(decrypt can also be added later)

Reply:
> (besides, should we later add benchmark for byte-array statistics decoding ?)

We can add benchmarks for whatever may be a performance bottleneck :-)

Reply:
The stats are already encoded so I think it is pretty trivial. But may be different if it is BYTE_ARRAY type with large strings.

Reply:
And we can add it in further patches, this could be a good start

Reply:
> The stats are already encoded so I think it is pretty trivial. But may be different if it is BYTE_ARRAY type with large strings.

That'd be valuable, but here we want to benchmark metadata cost in isolation of other factors?

Reply:
It's a good idea to also try this with the statistics dropped since these do have a big impact and have been heavily discussed. That is to say, run the numbers with `builder.disable_statistics()`. Based on previous profiling I have done in https://github.com/apache/arrow/issues/39676 this will cut the metadata read time by about 1/3. Still the main conclusion that the scaling is _O(num_columns * num_row_groups)_ should still hold. @pitrou

Reply:
Also, thanks a bunch for going to the effort to create these benchmarks! I think they are very helpful for the discussion!